Getting Started with JavaScript (#12689)
WordPress/gutenberg: docs/designers-developers/developers/tutorials/javascript/loading-javascript.md

# Loading JavaScript

With the plugin in place, you can add the code that loads the JavaScript. This methodology follows the standard WordPress procedure of enqueuing scripts, see [enqueuing section of the Plugin Handbook](https://developer.wordpress.org/plugins/javascript/enqueuing/).

Add the following code to your `myguten-plugin.php` file:

```php
function myguten_enqueue() {

> 	wp_enqueue_script( 'myguten-script',

Reviewer: I think per core code style this should be

```
wp_enqueue_script(
    'my-guten-script',
```

		plugins_url( 'myguten.js', __FILE__ )
	);
}
add_action( 'enqueue_block_editor_assets', 'myguten_enqueue' );
```

The `enqueue_block_editor_assets` hook is used, which is called when the block editor loads, and will enqueue the JavaScript file `myguten.js`.

Create a file called `myguten.js` and add:

```js
console.log( "I'm loaded!" );
```

> Now start a new post in the block editor, and check your browser Developer Tools. You should see the message in your console log. See [Mozilla's What are browser developer tools?](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools) if you need more information, the area to become most familiar with is [The JavaScript console](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools#The_JavaScript_console).

Reviewer: ```suggestion
Next, create a new post in the block editor.

We'll check the JavaScript console in your browser's Developer Tools, to see if the message is displayed. If you're not sure what developer tools are, Mozilla's ["What are browser developer tools?"](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools) documentation provides more information, including more background on the [JavaScript console](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools#The_JavaScript_console).

If your code is registered and enqueued correctly, you should see a message in your console:
```

_Then it would be nice to show a screenshot here of the successful console message._

> **Note for Theme Developers:**  The above method of enqueing is used for plugins, if you are extending the Block Editor for your theme there is a minor difference, you will use the `get_template_directory_uri()` function instead of `plugins_url()`. So for a theme, the enqueue example is:

Reviewer: ```suggestion
**Note for Theme Developers:**  The above method of enqueing is used for plugins. If you are extending the Block Editor for your theme there is a minor difference, you will use the `get_template_directory_uri()` function instead of `plugins_url()`. So for a theme, the enqueue example is:
```

```php
function myguten_enqueue() {

> 	wp_enqueue_script( 'myguten-script',

Reviewer: same code formatting note as above

		get_template_directory_uri() . '/myguten.js'
	);
}
add_action( 'enqueue_block_editor_assets', 'myguten_enqueue' );
```

### Recap

> At this point, you have a plugin in the directory `wp-content/plugins/myguten-plugin` with two files. The PHP server-side code in `myguten-plugin.php`, and the JavaScript which runs in the browser in `myguten.js`.

Reviewer: ```suggestion
At this point, you have a plugin in the directory `wp-content/plugins/myguten-plugin` with two files: the PHP server-side code in `myguten-plugin.php`, and the JavaScript which runs in the browser in `myguten.js`.
```

## Loading JavaScript

With the plugin in place, you can add the code that loads the JavaScript. This methodology follows the standard WordPress procedure of enqueuing scripts, see [enqueuing section of the Plugin Handbook](https://developer.wordpress.org/plugins/javascript/enqueuing/).

Add the following code to your `myguten-plugin.php` file:

```php
function myguten_enqueue() {
	wp_enqueue_script( 'myguten-script',
		plugins_url( 'myguten.js', __FILE__ ),

> 		array( 'wp-blocks', 'wp-element' )

Reviewer 2: I dig this step-by-step approach! It's very clear.

For the sake of creating a _Minimum Viable Plugin_, how about removing these dependencies. They're not required just yet to write to the console.

	);
}
add_action( 'enqueue_block_editor_assets', 'myguten_enqueue' );
```

Create a file called `myguten.js` and add:

```js
console.log( "I'm loaded!" );
```

Now start a new post in the block editor, and check your browser Developer Tools. You should see the message in your console log. See [Mozilla's What are browser developer tools?](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools) if you need more information, the area to become most familiar with is [The JavaScript console](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools#The_JavaScript_console).

The script enqueuing uses the `enqueue_block_editor_assets` hook which only loads the JavaScript when the block editor loads. So if you navigate to any front-end page or article on your site, you will not see the message.

> The last argument in the `wp_enqueue_script()` function is an array of dependencies. All of the Gutenberg packages are registered and can be loaded by specifying them in the array, the blocks and elements packages are shown above, as they are two common ones used.

Reviewer 2: We could remove this part about dependencies and add it back in the first example that extends the editor.

Author: I rearranged the enqueue with two steps explaining it more clearly in the extension part.

> **Note for Theme Developers:**  The above method of enqueing is used for plugins, if you are extending Gutenberg for your theme there is a minor difference, you will use the `get_template_directory_uri()` function instead of `plugins_url()`. So for a theme, the enqueue example is:

Reviewer 2: :heart:

```php
function myguten_enqueue() {
	wp_enqueue_script( 'myguten-script',
		get_template_directory_uri() . '/myguten.js',

> 		array( 'wp-blocks', 'wp-element' )

Reviewer 2: Same comment here about dependencies.

If your code is registered and enqueued correctly, you should see a message in your console:

> ![Console Log Message Success](../../../../../docs/designers-developers/assets/js-tutorial-console-log-success.png)

Reviewer: With image assets I think we're cool to make them full URLs (and in fact I believe we have to right now; the handbook won't sync images and I think needs the raw URLs linked here)

Author: so the URL from Github, like so?

`https://raw.githubusercontent.com/WordPress/gutenberg/docs/11392/javascript/docs/designers-developers/assets/js-tutorial-console-log-error.png`

The above includes the branch, so changes after commit

Reviewer: Yep, you can see how we do it here: https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/designers-developers/designers/block-design.md

Author: Updated. 

The images won't show in PR view, since they aren't in master until after merge.
